table/block_based_table_reader.cc

```diff
@@ -173,6 +173,25 @@ Cache::Handle* GetEntryFromCache(Cache* block_cache, const Slice& key,
   return cache_handle;
 }
 
+bool PrefixExtractorChanged(std::string prefix_extractor_block,
+                            const SliceTransform* prefix_extractor,
+                            BlockBasedTableOptions::IndexType index_type) {
+  bool prefix_extractor_changed = false;
+  // only needed for hash based index
+  if (index_type != BlockBasedTableOptions::kHashSearch &&
+      prefix_extractor != nullptr) {
+    return prefix_extractor_changed;
+  }
+  if (prefix_extractor_block.compare("nullptr") != 0 && prefix_extractor != nullptr &&
+      prefix_extractor_block.compare(prefix_extractor->Name()) != 0) {
+    prefix_extractor_changed = true;
+  }
+  else if (prefix_extractor_block.compare("nullptr") != 0 || prefix_extractor != nullptr) {
+    prefix_extractor_changed = true;
+  }
+  return prefix_extractor_changed;
```

Comment:
It feels like there are too many string comparisons done for the most common case, which is the prefix extractor didn't change. Please make sure that only one comparison is done in the case of `prefix_extractor_block.compare(prefix_extractor->Name()) == 0`